kvserver, storage: clean up, fix bugs in, and test snapshot receive path

tbg opened the pull request:
When a snapshot is received, we write the incoming KV pairs into a number of SSTs. With the introduction of excise and shared/external SSTs (disagg storage), this code had become quite crufty and supported - in principle - multiple modes that made it quite brittle and under-tested.

In this PR, we "bake in" excises: for the SST covering the MVCC span, we don't lay down a range deletion but instead *always* use excise. We now also always allow splitting of large MVCC SSTs (into reasonably-sized SSTs). This works against the previous proliferation of options, and enabled various code improvements.

The code around range keys (MVCC rangedels) was cleaned up too, and missing fragmentation was added. For "regular" rangedels (ClearRange), fragmentation was missing entirely and was added as well.

We now have a datadriven unit test for the MultiSSTWriter, which shows precisely what's contained in each produced SST. This doesn't mean the receive path is well-tested - far from it - but at least there's a solid foundation to build on top of.

Overall, the code here is much improved and easier to reason about.

Epic: CRDB-46488
Release Note: None

tbg committed:
kvserver: leave SST splitting and fragmenter enabled

I believe the only reason for `!skipClearForMVCCSpan` disabling MVCC SST
splitting was that in previous versions of the code, the needed to anticipate a
call to `addClearForMVCCSpan()`. I don't quite understand why that would really
have presented a need to disable splitting - after all, that call would have
occurred before even a single key had been added to the MVCC span - but my best
explanation is that this wasn't always clear to all involved authors of the
code. This echoes in the range key fragmenter being bypassed in this mode as
well.

At this point, we don't actually need to support addClearForMVCCSpan (it was
removed in #142651), and we unconditionally set `skipClearForMVCCSpan`.

We would like to enjoy the benefits of SST splitting, so this commit enables
it, which requires us to also fragment range keys correctly. I believe we
previously bypassed the fragmenter because we worried that
`addClearForMVCCSpan` would need to have these rangekeys in the SST already.
Though I don't think in practice we supported that - the SST and fragmenter
would've been entirely empty - see above and the second commit in #142476.

Epic: CRDB-46488

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/142653)
<!-- Reviewable:end -->

tbg committed:
kvserver: simplify finalizeSST

In the MVCC span, we may deal with SSTs splitting MVCC rows between timestamps, so we need to use EngineKey, whereas in the local key spans we previously dealt in `roachpb.Key`. We were calling slightly different versions of clearrange.

This commit unifies the code paths, making it more transparent what's going on.

tbg committed:
kvserver: set skipClearForMVCCSpan to true

This required a little change in `msstw.PutRangeKey`.

That method was assuming that range keys only occur in the MVCC span, which is
true in production, but is not something we want to rely on. In particular,
`TestRaftSnapshotsWithMVCCRangeKeysEverywhere` exercises the case of a snapshot
which has a range key in the range-id replicated span.

tbg committed:
kvserver: remove skipClearForMVCCSpan

It was hard-coded to true anyway.

Adding a comment while was there. It's quite easy to get confused between all
the range key and clear range operations.

tbg committed:
kvserver: move snap sst rangedel to init time

The only reason to clear at SST finish time was that we (in the past) sometimes
wanted to clear the MVCC span as well, and that span can split across SSTs,
meaning the bounds weren't known a priori. This is no longer a concern, since
excise is used on the MVCC span. As a result, we get to simplify by putting
both kinds of deletions in a single place (though technically speaking due
to the use of the fragmenter for the range key deletion, it "takes effect"
only at the end).

tbg committed:
kvserver: improve msstw fragmenter comment

tbg committed:
kvserver: remove includesRangeDelForLastSpan

tbg committed:
storage: remove unused param to IngestAndExciseFiles

RaduBerinde reviewed (commented):
Very nice cleanup, thanks! I have just one question about fragmenting range dels.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142653#-:-OLGDHu67wDUtDKiIzkK:b-1sjert)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @jbowens)

___
*[`pkg/kv/kvserver/multi_sst_writer.go` line 59 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/142653#-OLGD1mS5YH9LCSXBgZM:-OLGD1mS5YH9LCSXBgZN:b9c288c) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ed919e986edeb49d7a34322357f1315d597b1ed/pkg/kv/kvserver/multi_sst_writer.go#L59)):*
> ```Go
> 	// For each SST, it's initialized with a range key del for the entire span,
> 	// but the incoming stream of data may also contain new range keys. The
> 	// fragmenter is emits these range keys into the SST at finalization
> ```

[nit] remove "is"
___
*[`pkg/kv/kvserver/multi_sst_writer.go` line 146 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/142653#-OLGCojlDoQ67x8NdGtw:-OLGCojlDoQ67x8NdGtx:b-qdmubt) ([raw file](https://github.com/cockroachdb/cockroach/blob/4f2e84bc68a14ecd0c394d1e777becb017ab060a/pkg/kv/kvserver/multi_sst_writer.go#L146)):*
> ```Go
> 		s := rangekey.Span{Start: startKey.Encode(), End: endKey.Encode(), Keys: []rangekey.Key{{Trailer: trailer}}}
> 		msstw.rangeKeyFrag.Add(s)
> ```

Don't we need a fragmenter for the range dels as well? I don't see why they should differ in this respect
___
*[`pkg/kv/kvserver/replica_sst_snapshot_storage_test.go` line 475 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/142653#-OLGCU8d2Jh_UyEu0Ybw:-OLGCU8d2Jh_UyEu0Ybx:b-4vqrfb) ([raw file](https://github.com/cockroachdb/cockroach/blob/9235f5c679dc448d049759bcb4f6a59fe3f99db9/pkg/kv/kvserver/replica_sst_snapshot_storage_test.go#L475)):*
> ```Go
> 	mvccSpan := keySpans[len(keySpans)-1]
> 
> 	msstw, err := newMultiSSTWriter(ctx, cluster.MakeTestingClusterSettings(), scratch, localSpans, mvccSpan, 0, false)
> ```

[nit] add comment with arg name
___
*[`pkg/kv/kvserver/client_merge_test.go` line 3912 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/142653#-OLGBpL_6hpL2JWaLhPJ:-OLGBpL_6hpL2JWaLhPK:b-math8y) ([raw file](https://github.com/cockroachdb/cockroach/blob/d18b04f78a6b9d04c35fb9239c4b5db7cd802f1d/pkg/kv/kvserver/client_merge_test.go#L3912)):*
> ```Go
> 			file := &storage.MemObject{}
> 			writer := storage.MakeIngestionSSTWriter(ctx, st, file)
> 			if i < len(keySpans)-1 {
> ```

[nit] Add a comment explaining this


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142653)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @jbowens)

___
*[`pkg/kv/kvserver/multi_sst_writer.go` line 146 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/142653#-OLGCojlDoQ67x8NdGtw:-OLGIqhMDNvmWAnGYKiC:btspm7e) ([raw file](https://github.com/cockroachdb/cockroach/blob/4f2e84bc68a14ecd0c394d1e777becb017ab060a/pkg/kv/kvserver/multi_sst_writer.go#L146)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Don't we need a fragmenter for the range dels as well? I don't see why they should differ in this respect
</blockquote></details>

That's a good point, this seems broken as far as I can tell. If we're receiving a shared SST, it can contain rangedels and these would be passed to the writer. The writer expects fragmented rangedels, so we do need to fragment the one we add into the incoming ones.

https://github.com/cockroachdb/cockroach/blob/0775dd4e4e764979c633403f417c67221d5d7d48/pkg/kv/kvserver/kv_snapshot_strategy.go#L288-L359

Taking a step back though, I also thought shared SSTs didn't even get streamed as write batches. So I don't actually understand why `readOneToBatch` handles key kinds that can only exist in a shared SST - for shared SSTs, we just receive metadata that we (later) pass on to the ingestion:

https://github.com/cockroachdb/cockroach/blob/0775dd4e4e764979c633403f417c67221d5d7d48/pkg/kv/kvserver/kv_snapshot_strategy.go#L201-L217


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
*[`pkg/kv/kvserver/multi_sst_writer.go` line 146 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/142653#-OLGCojlDoQ67x8NdGtw:-OLH8W4J1NVqlrnSZZLn:b-wc11yi) ([raw file](https://github.com/cockroachdb/cockroach/blob/4f2e84bc68a14ecd0c394d1e777becb017ab060a/pkg/kv/kvserver/multi_sst_writer.go#L146)):*
> Taking a step back though, I also thought shared SSTs didn't even get streamed as write batches

Good question. I checked and even for shared SSTs that only partially overlap the target key span, we mess with their bounds and still skip them.


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142653)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @jbowens and @tbg)

___
*[`pkg/kv/kvserver/multi_sst_writer.go` line 146 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/142653#-OLGCojlDoQ67x8NdGtw:-OLHBNTx1DlHz1-EDSgW:b-go9y0n) ([raw file](https://github.com/cockroachdb/cockroach/blob/4f2e84bc68a14ecd0c394d1e777becb017ab060a/pkg/kv/kvserver/multi_sst_writer.go#L146)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

> Taking a step back though, I also thought shared SSTs didn't even get streamed as write batches

Good question. I checked and even for shared SSTs that only partially overlap the target key span, we mess with their bounds and still skip them.
</blockquote></details>

Here's a draft PR which adds a panic for these kinds, let's see what fails if anything: https://github.com/cockroachdb/cockroach/pull/142876


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
*[`pkg/kv/kvserver/multi_sst_writer.go` line 146 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/142653#-OLGCojlDoQ67x8NdGtw:-OLHQCLF8PhQ32rYTgkz:b-hsam9f) ([raw file](https://github.com/cockroachdb/cockroach/blob/4f2e84bc68a14ecd0c394d1e777becb017ab060a/pkg/kv/kvserver/multi_sst_writer.go#L146)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Here's a draft PR which adds a panic for these kinds, let's see what fails if anything: https://github.com/cockroachdb/cockroach/pull/142876
</blockquote></details>

Nothing fails, apparently, we should just remove these. They may have been temporarily needed at some point before the skip-shared implementation was complete.


<!-- Sent from Reviewable.io -->

tbg committed:
kvserver: demonstrate lack of rangedel fragmentation

This should work, but (if uncommented) crashes.
We need to add rangedel fragmentation to multiSSTWriter.
This will be done in a follow-up commit.

tbg committed:
kvserver: add rangedel fragmentation

Note that the rangedel now added in the test is a little crazy,
due to the unwise use of `.Next()`. This will be fixed up in
follow-up commits to make the test more realistic and less
confusing.

tbg committed:
kvserver: add a comment about potential bug

tbg committed:
kvserver: improve and test msstw rangekey fragmentation

This now fragments correctly in PutInternalRangeKey, which I think is where
this was actually needed. I consolidated the impl of `PutRangeKey` to call
`PutInternalRangeKey` to avoid further discrepancies.

TestMultiSSTWriterInitSST has been updated with range key sets on all key spans
to make sure that the fragmentation works.

While working on this I noticed a quirk: `sst.EstimatedSize()` does not change
when range keys are added. This is because SSTWriter has an internal fragmenter
as well, and keys in the fragmenter don't change EstimatedSize().  The double
layering of fragmenters is needed: we need to fragment our keys over multiple
SSTs).

Presumably, if we had lots of range keys, the start key would move forward and
we'd flush a substantial portion of the keys. So this inaccuracy is likely not
going to cause any issues.

tbg commented on pkg/kv/kvserver/testdata/echotest/TestMultiSSTWriterInitSST:
```diff
@@ -0,0 +1,29 @@
+>> sst4:
+rangedel: {d-/Max}
+rangekeyset: /Local/Lock/{Max-""\x12\xff\xff\x00\x01\x00""} -> /<empty>
+rangekeyset: {d\x00-\xff\xff\x00} -> /<empty>
```
This seems fishy. Why the `\x00` on both start and end key?

RaduBerinde replied:
Oh, in this case these are always just prefixes, but all valid prefixes end in \x00 (they are also valid keys).

tbg replied:
There was a bug here :-) I doubly encoded keys in some case (see [here](https://docs.google.com/document/d/1uvyFWuJg4FWl2OVwg3ekQlVNBb5CjCj_muUr3HG-oKU/edit?tab=t.0))

I fixed it now, the \x00 are gone, and CI passes. 😅

RaduBerinde replied:
Oh, sorry, then this must print the roach key not the encoded prefix (i.e. roach key + \x00).

tbg committed:
testcluster: guard against ChangeReplica on non-start key

Cost me a few minutes to figure out that replication changes need to specify an exact start key.
I figured they would look up the range descriptor first and then use the start key, but this is
not the case.

tbg commented:
bors r-

Waiting out the branch cut. I wouldn't be entirely surprised if this PR causes some surprises.

pav-kv commented on pkg/kv/kvserver/multi_sst_writer.go:
```diff
@@ -406,12 +403,8 @@ func (msstw *multiSSTWriter) PutRangeKey(
-		Keys:  []rangekey.Key{{Trailer: startTrailer, Suffix: suffix, Value: value}},
-	})
-	return nil
+	rk := rangekey.Key{Trailer: startTrailer, Suffix: suffix, Value: value}
+	return msstw.PutInternalRangeKey(ctx, startKey, endKey, rk)
```
There is some redundancy after this change: we encode the keys, and they are decoded back in `PutInternalRangeKey`. Also, `rolloverSST` is called twice - is this legit?

My preference would be to either denormalize this code again, or structure it so that there is no redundancy.

pav-kv reviewed (commented):
First pass. Looks good, just some nits (all optional) and questions. Also a couple of pseudo-bugs.

pav-kv commented on pkg/kv/kvserver/testdata/echotest/TestMultiSSTWriterInitSST:
```diff
@@ -1,20 +1,21 @@
-rangedel: /Local/Lock/{""\x12d\x00\x01\x00""-Max}
+rangedel: /Local/Lock{"d"-/Max}
+rangekeyset: /Local/Lock{"d"-/Max} -> /<empty>
 rangekeydel: /Local/Lock{"d"-/Max}
```
What does it mean that these two `rangekey` things overlap by the span? What is the precedence between them?

tbg replied:
My understanding is that rangedels and rangekeys are different. Rangedels clear points in their span, whereas rangekey are MVCC deletion markers for a key span. Furthermore, range keys can "stack", since you can have overlapping ones at different timestamps (the timestamps aren't really visible here because I'm not putting any). Due to fragmentation, we end up with exact overlapping copies.

Btw, I'm not sure I see your comment in the right place, but these are the general thoughts that hopefully explain what you were looking at.

tbg commented on pkg/kv/kvserver/client_merge_test.go:
```diff
@@ -3905,11 +3905,13 @@ func TestStoreRangeMergeRaftSnapshot(t *testing.T) {
 			writer := storage.MakeIngestionSSTWriter(ctx, st, file)
-			if err := writer.ClearRawRange(span.Key, span.EndKey, true /* pointKeys */, true /* rangeKeys */); err != nil {
-				return err
+			if i < len(keySpans)-1 {
```
Adding:

				// The last span is the MVCC span, and is always cleared via Excise.
				// See multiSSTWriter.

tbg committed:
kvserver: add a test comment

tbg commented on pkg/kv/kvserver/multi_sst_writer.go:
```diff
@@ -54,6 +54,10 @@ type multiSSTWriter struct {
 	maxSSTSize int64
 	// rangeKeyFrag is used to fragment range keys across the mvcc key spans.
+	// For each SST, it's initialized with a range key del for the entire span,
+	// but the incoming stream of data may also contain new range keys. The
+	// fragmenter is emits these range keys into the SST at finalization
+	// time.
```
Your understanding is correct, I improved the comment:

	// rangeKeyFrag is used to fragment range keys across the SSTs. For each SST
	// (with the exception of the MVCC SST), it's initialized with a range key del
	// for the entire span, but the incoming stream of data may also contain new
	// range keys. As the start key of these incoming range keys increases, the
	// fragmenter emits fragmented range keys into the produced SST.

tbg committed:
kvserver: improve fragmenter comments

tbg committed:
storage: multi-line formatting for IngestAndExciseFiles

tbg committed:
kvserver: account for writeBytes in flushed fragments

The previous accounting was a no-op, since we weren't adding directly to the
SSTs.

tbg committed:
kvserver: make TestMultiSSTWriter more interesting

tbg committed:
kvserver: further normalize PutInternalRangeDelete

This also fixes a buglet, namely calling rolloverSST twice.
The second call would be a no-op, so this wasn't a consequential bug.

pav-kv commented on pkg/kv/kvserver/multi_sst_writer.go:
```diff
@@ -382,17 +382,43 @@ func (msstw *multiSSTWriter) PutInternalRangeDelete(ctx context.Context, start,
+	return msstw.putRangeKey(ctx, storage.EngineKeyRange{}, [2][]byte{start, end}, key)
+}
+
+func (msstw *multiSSTWriter) putRangeKey(
```
I have mixed feelings about this one. The amount of duplication in the old code wasn't bad, only needed to call `rolloverSST` and `Frag.Add`. The `rolloverSST` is done by all other similar funcs, so it's not a bad duplication, just following the pattern. So only `Frag.Add()` would be "badly" duplicated. Seems ok, if compared to extra complexity in this normalization commit.

Alternatively, maybe this is fixable further down the stack, so that e.g. `rolloverSST` takes a different type. Haven't thought about this much.

tbg replied:
Yeah, you're right, the reason for deduplicating originally was that a bunch of code was duplicated (and the important code was missing in the copy that mattered), but with the movement of stats into the fragmenters it's less the case now. I'll stop playing code golf here for now, though, it's clear what the code does and it's as obviously correct as maintaining two versions of the code. Could've ended up the other way in a parallel universe.

pav-kv reviewed (approved):
LGTM. Outstanding comments not blocking - just readability stuff.

pav-kv commented on pkg/kv/kvserver/testdata/echotest/TestMultiSSTWriterInitSST_interesting=true:
```diff
@@ -0,0 +1,37 @@
+rangekeydel: /Local/RangeID/100/r/AbortSpan/"{00000000-0000-0000-0000-000000000000"-6ba7b810-9dad-11d1-80b4-00c04fd430c8"}
+rangekeydel: /Local/RangeID/100/{r/AbortSpan/"6ba7b810-9dad-11d1-80b4-00c04fd430c8"-s""}
+>> sst1:
+rangedel: /Local/Range"d"{-/RangeDescriptor}
+rangedel: /Local/Range"{d"/RangeDescriptor-f"/RangeDescriptor}
+rangedel: /Local/Range{"f"/RangeDescriptor-/Max}
```
I'm wondering: why are we trying hard to fragment rangedels if the end effect is still that the entire span is cleared? Could we ignore rangedels instead of fragmenting them?

tbg replied:
I'm not sure, this is a bit above my paygrade. I _think_ each kv pair in an SST has its own seqno. So it is not semantically correct to "merge" rangedels at different sequence numbers. Of course this can't be quite true here, since the SSTs are only ingested later, which is when their seqnos are assigned (not sure how this works exactly).

Either way, what the code path does here is essentially "replay" a WAL from another node on top of a base SST (from shared storage). I don't think we should start doing anything special here. It matters that the resulting SSTs are well-formed, and the fragmenter makes sure this is the case, under no assumption on the input data. I don't feel a need to push it past that point, but let me ask the storage folks about how this works so that we can understand better what exactly happens here.

tbg replied:
Re: second comment, yes, it can be.

tbg replied:
https://cockroachlabs.slack.com/archives/CAC6K3SLU/p1744383586300569

tbg commented:
bors r+
